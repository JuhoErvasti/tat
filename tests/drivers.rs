use tat::{fixtures::{datasets::*, init_table, TatTestStructure, TatTestUtils}, table::TatTable};
use rstest::rstest;
use insta::assert_snapshot;

fn test_single_layer_dataset(driver_name: &str, structs: (TatTestStructure, TatTable)) {
    let (test, mut t) = structs;
    TatTestUtils::set_layer_index_and_update(0, &mut t, &test.tatevent_rx);
    TatTestUtils::prepare_table_attribute_view_all_features(&t, &test.tatevent_rx, test.ds_request_tx.clone());

    let snapshot_name = format!("{}_{}", driver_name, t.layer_schema().unwrap().name());
    assert_snapshot!(snapshot_name, t);
}

fn test_all_layers(driver_name: &str, structs: (TatTestStructure, TatTable)) {
    let (test, mut t) = structs;
    for index in 0..t.layer_schemas().len() as usize {
        TatTestUtils::set_layer_index_and_update(index, &mut t, &test.tatevent_rx);
        TatTestUtils::prepare_table_attribute_view_all_features(&t, &test.tatevent_rx, test.ds_request_tx.clone());

        assert_eq!(t.layer_schema().unwrap().index(), index);
        let snapshot_name = format!("{}_{}", driver_name, t.layer_schema().unwrap().name());
        assert_snapshot!(snapshot_name, t);
    }

    test.terminate();
}

#[rstest]
fn test_basic_gdb(basic_gdb: TatTestStructure) {
    test_all_layers("openfilegdb", init_table(basic_gdb));
}

#[rstest]
fn test_basic_gml(basic_gml: TatTestStructure) {
    test_all_layers("gml", init_table(basic_gml));
}

#[rstest]
fn test_basic_jsonfg(basic_jsonfg: TatTestStructure) {
    test_all_layers("jsonfg", init_table(basic_jsonfg));
}

#[rstest]
fn test_basic_mapml(basic_mapml: TatTestStructure) {
    test_all_layers("mapml", init_table(basic_mapml));
}

#[rstest]
fn test_basic_ods(basic_ods: TatTestStructure) {
    test_all_layers("ods", init_table(basic_ods));
}

// the reason these are individual functions like this is because it works
// better with the insta snapshots
#[rstest]
fn test_basic_shp_point(basic_shp_point: TatTestStructure) {
    test_single_layer_dataset("esri_shapefile", init_table(basic_shp_point));
}

#[rstest]
fn test_basic_shp_line(basic_shp_line: TatTestStructure) {
    test_single_layer_dataset("esri_shapefile", init_table(basic_shp_line));
}

#[rstest]
fn test_basic_shp_polygon(basic_shp_polygon: TatTestStructure) {
    test_single_layer_dataset("esri_shapefile", init_table(basic_shp_polygon));
}

#[rstest]
fn test_basic_shp_multipolygon(basic_shp_multipolygon: TatTestStructure) {
    test_single_layer_dataset("esri_shapefile", init_table(basic_shp_multipolygon));
}

#[rstest]
fn test_basic_shp_nogeom(basic_shp_nogeom: TatTestStructure) {
    test_single_layer_dataset("esri_shapefile", init_table(basic_shp_nogeom));
}

#[rstest]
fn test_basic_csv_point(basic_csv_point: TatTestStructure) {
    test_single_layer_dataset("csv", init_table(basic_csv_point));
}

#[rstest]
fn test_basic_csv_line(basic_csv_line: TatTestStructure) {
    test_single_layer_dataset("csv", init_table(basic_csv_line));
}

#[rstest]
fn test_basic_csv_polygon(basic_csv_polygon: TatTestStructure) {
    test_single_layer_dataset("csv", init_table(basic_csv_polygon));
}

#[rstest]
fn test_basic_csv_multipolygon(basic_csv_multipolygon: TatTestStructure) {
    test_single_layer_dataset("csv", init_table(basic_csv_multipolygon));
}

#[rstest]
fn test_basic_csv_nogeom(basic_csv_nogeom: TatTestStructure) {
    test_single_layer_dataset("csv", init_table(basic_csv_nogeom));
}

#[rstest]
fn test_basic_geojson_point(basic_geojson_point: TatTestStructure) {
    test_single_layer_dataset("geojson", init_table(basic_geojson_point));
}

#[rstest]
fn test_basic_geojson_line(basic_geojson_line: TatTestStructure) {
    test_single_layer_dataset("geojson", init_table(basic_geojson_line));
}

#[rstest]
fn test_basic_geojson_polygon(basic_geojson_polygon: TatTestStructure) {
    test_single_layer_dataset("geojson", init_table(basic_geojson_polygon));
}

#[rstest]
fn test_basic_geojson_multipolygon(basic_geojson_multipolygon: TatTestStructure) {
    test_single_layer_dataset("geojson", init_table(basic_geojson_multipolygon));
}

#[rstest]
fn test_basic_geojson_nogeom(basic_geojson_nogeom: TatTestStructure) {
    test_single_layer_dataset("geojson", init_table(basic_geojson_nogeom));
}

#[rstest]
fn test_basic_geojsonseq_point(basic_geojsonseq_point: TatTestStructure) {
    test_single_layer_dataset("geojsonseq", init_table(basic_geojsonseq_point));
}

#[rstest]
fn test_basic_geojsonseq_line(basic_geojsonseq_line: TatTestStructure) {
    test_single_layer_dataset("geojsonseq", init_table(basic_geojsonseq_line));
}

#[rstest]
fn test_basic_geojsonseq_polygon(basic_geojsonseq_polygon: TatTestStructure) {
    test_single_layer_dataset("geojsonseq", init_table(basic_geojsonseq_polygon));
}

#[rstest]
fn test_basic_geojsonseq_multipolygon(basic_geojsonseq_multipolygon: TatTestStructure) {
    test_single_layer_dataset("geojsonseq", init_table(basic_geojsonseq_multipolygon));
}

#[rstest]
fn test_basic_geojsonseq_nogeom(basic_geojsonseq_nogeom: TatTestStructure) {
    test_single_layer_dataset("geojsonseq", init_table(basic_geojsonseq_nogeom));
}

#[rstest]
fn test_basic_jml_point(basic_jml_point: TatTestStructure) {
    test_single_layer_dataset("jml", init_table(basic_jml_point));
}

#[rstest]
fn test_basic_jml_line(basic_jml_line: TatTestStructure) {
    test_single_layer_dataset("jml", init_table(basic_jml_line));
}

#[rstest]
fn test_basic_jml_polygon(basic_jml_polygon: TatTestStructure) {
    test_single_layer_dataset("jml", init_table(basic_jml_polygon));
}

#[rstest]
fn test_basic_jml_multipolygon(basic_jml_multipolygon: TatTestStructure) {
    test_single_layer_dataset("jml", init_table(basic_jml_multipolygon));
}

#[rstest]
fn test_basic_jml_nogeom(basic_jml_nogeom: TatTestStructure) {
    test_single_layer_dataset("jml", init_table(basic_jml_nogeom));
}

#[rstest]
fn test_basic_mapinfofile_point(basic_mapinfofile_point: TatTestStructure) {
    test_single_layer_dataset("mapinfo_file", init_table(basic_mapinfofile_point));
}

#[rstest]
fn test_basic_mapinfofile_line(basic_mapinfofile_line: TatTestStructure) {
    test_single_layer_dataset("mapinfo_file", init_table(basic_mapinfofile_line));
}

#[rstest]
fn test_basic_mapinfofile_polygon(basic_mapinfofile_polygon: TatTestStructure) {
    test_single_layer_dataset("mapinfo_file", init_table(basic_mapinfofile_polygon));
}

#[rstest]
fn test_basic_mapinfofile_multipolygon(basic_mapinfofile_multipolygon: TatTestStructure) {
    test_single_layer_dataset("mapinfo_file", init_table(basic_mapinfofile_multipolygon));
}

#[rstest]
fn test_basic_mapinfofile_nogeom(basic_mapinfofile_nogeom: TatTestStructure) {
    test_single_layer_dataset("mapinfo_file", init_table(basic_mapinfofile_nogeom));
}

#[rstest]
fn test_basic_xlsx_point(basic_xlsx_point: TatTestStructure) {
    test_single_layer_dataset("xlsx", init_table(basic_xlsx_point));
}

#[rstest]
fn test_basic_xlsx_line(basic_xlsx_line: TatTestStructure) {
    test_single_layer_dataset("xlsx", init_table(basic_xlsx_line));
}

#[rstest]
fn test_basic_xlsx_polygon(basic_xlsx_polygon: TatTestStructure) {
    test_single_layer_dataset("xlsx", init_table(basic_xlsx_polygon));
}

#[rstest]
fn test_basic_xlsx_multipolygon(basic_xlsx_multipolygon: TatTestStructure) {
    test_single_layer_dataset("xlsx", init_table(basic_xlsx_multipolygon));
}

#[rstest]
fn test_basic_xlsx_nogeom(basic_xlsx_nogeom: TatTestStructure) {
    test_single_layer_dataset("xlsx", init_table(basic_xlsx_nogeom));
}
