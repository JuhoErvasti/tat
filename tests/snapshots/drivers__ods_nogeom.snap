---
source: tests/drivers.rs
expression: result
---
text_field,i32_field,i64_field,decimal_field,date_field,datetime_field,bool_field,blob_field,json_field
text,10,100,1.541,1970/07/10,2025/07/19 20:45:45,1,626C6F620A,{"another_key":"another_value","key":"value"}
buzz,374,-1384,-104.092138,2010/05/25,2016/05/03 01:40:15.048,0,626C6F620A,{"another_key":"another_value","key":"value"}
check,173,2474,-170.296267,2009/02/22,2015/12/13 04:04:17.556,0,626C6F620A,{"another_key":"another_value","key":"value"}
die,681,-404,7.189099,2028/04/06,2023/12/28 17:57:16.036,1,626C6F620A,{"another_key":"another_value","key":"value"}
participate,967,-3170,72.283053,2022/02/13,2025/12/12 02:11:45.662,0,626C6F620A,{"another_key":"another_value","key":"value"}
huge,980,-403,90.188894,2013/06/29,2018/01/30 07:39:08.074,0,626C6F620A,{"another_key":"another_value","key":"value"}
read,295,-1531,-84.884592,2023/08/27,2022/01/13 19:28:32.476,0,626C6F620A,{"another_key":"another_value","key":"value"}
check,894,301,165.484085,2024/06/07,2024/03/09 14:32:41.238,0,626C6F620A,{"another_key":"another_value","key":"value"}
strive,841,-1536,43.659255,2020/12/29,2029/01/26 07:53:25.466,0,626C6F620A,{"another_key":"another_value","key":"value"}
strive,729,4101,-192.625146,2018/12/22,2019/12/23 21:51:52.335,0,626C6F620A,{"another_key":"another_value","key":"value"}
fall,456,3834,-152.688532,2029/04/17,2016/07/04 06:13:35.176,1,626C6F620A,{"another_key":"another_value","key":"value"}
participate,595,-219,56.665858,2016/06/23,2009/09/10 01:26:32.349,1,626C6F620A,{"another_key":"another_value","key":"value"}
study,518,4042,14.859642,2028/12/22,2015/03/22 08:33:33.770,1,626C6F620A,{"another_key":"another_value","key":"value"}
respect,577,3366,5.526759,2012/11/27,2019/01/02 10:08:57.047,1,626C6F620A,{"another_key":"another_value","key":"value"}
die,434,2946,164.085165,2025/06/11,2026/06/17 07:23:28.466,1,626C6F620A,{"another_key":"another_value","key":"value"}
verify,-28,1443,45.941403,2027/03/03,2008/06/10 19:20:09.165,1,626C6F620A,{"another_key":"another_value","key":"value"}
study,206,2529,-129.470353,2009/06/05,2009/12/15 09:45:58.501,1,626C6F620A,{"another_key":"another_value","key":"value"}
verify,540,-3621,70.006604,2026/02/26,2022/01/21 08:04:41.328,0,626C6F620A,{"another_key":"another_value","key":"value"}
verify,-92,2104,85.831181,2021/03/05,2022/04/05 02:29:18.756,1,626C6F620A,{"another_key":"another_value","key":"value"}
murky,163,3234,59.299899,2016/11/23,2012/07/28 01:56:33.320,0,626C6F620A,{"another_key":"another_value","key":"value"}
arrest,670,2865,-165.115901,2019/04/19,2026/11/27 03:46:43.114,1,626C6F620A,{"another_key":"another_value","key":"value"}
participate,261,204,-90.869299,2013/07/16,2019/11/13 17:56:37.309,0,626C6F620A,{"another_key":"another_value","key":"value"}
loving,615,4905,-135.931367,2024/06/12,2015/01/11 10:11:40.443,0,626C6F620A,{"another_key":"another_value","key":"value"}
breath,873,-3744,23.026825,2009/06/17,2021/06/18 19:26:26.856,1,626C6F620A,{"another_key":"another_value","key":"value"}
breath,507,-1789,181.181897,2026/02/08,2011/04/23 12:40:14.519,1,626C6F620A,{"another_key":"another_value","key":"value"}
read,134,307,-115.438984,2008/02/02,2029/01/05 19:17:11.004,0,626C6F620A,{"another_key":"another_value","key":"value"}
breath,177,-4535,-20.305131,2025/06/03,2017/05/24 17:41:15.977,0,626C6F620A,{"another_key":"another_value","key":"value"}
participate,400,2477,57.760561,2023/09/04,2009/11/14 14:39:00.661,1,626C6F620A,{"another_key":"another_value","key":"value"}
tremble,-7,-4141,163.773982,2013/02/20,2023/05/15 05:32:31.008,0,626C6F620A,{"another_key":"another_value","key":"value"}
study,873,-3609,45.920409,2011/07/05,2027/03/31 08:35:38.187,1,626C6F620A,{"another_key":"another_value","key":"value"}
raspy,842,4792,95.786027,2015/06/20,2021/05/14 10:08:54.558,0,626C6F620A,{"another_key":"another_value","key":"value"}
current,323,-1861,-82.259782,2019/09/02,2020/01/20 22:50:05.982,1,626C6F620A,{"another_key":"another_value","key":"value"}
bloody,418,4161,125.614895,2014/06/16,2019/01/07 06:42:48.012,1,626C6F620A,{"another_key":"another_value","key":"value"}
murky,918,-4092,196.367059,2021/06/24,2025/01/11 09:39:58.739,0,626C6F620A,{"another_key":"another_value","key":"value"}
current,212,4349,106.562075,2026/12/26,2011/06/16 13:44:14.665,1,626C6F620A,{"another_key":"another_value","key":"value"}
read,-51,-425,154.331481,2026/06/09,2013/10/12 12:47:54.145,1,626C6F620A,{"another_key":"another_value","key":"value"}
verify,870,-3350,73.29785,2013/01/29,2011/04/02 17:39:28.947,1,626C6F620A,{"another_key":"another_value","key":"value"}
loving,13,2095,98.944013,2017/09/20,2028/12/11 02:01:38.787,1,626C6F620A,{"another_key":"another_value","key":"value"}
verify,356,-4767,-103.234727,2008/08/18,2016/10/09 16:58:43.109,0,626C6F620A,{"another_key":"another_value","key":"value"}
die,3,-3577,-61.458799,2021/08/13,2023/01/07 15:59:55.057,0,626C6F620A,{"another_key":"another_value","key":"value"}
raspy,634,-113,-13.152387,2019/01/16,2027/04/28 21:35:52.459,0,626C6F620A,{"another_key":"another_value","key":"value"}
murky,306,3620,-1.47147,2021/11/25,2016/03/15 04:34:20.185,1,626C6F620A,{"another_key":"another_value","key":"value"}
bad,987,2034,15.920554,2009/09/08,2022/08/28 19:16:38.777,0,626C6F620A,{"another_key":"another_value","key":"value"}
courageous,584,-90,163.497059,2023/03/24,2010/05/29 07:45:13.766,0,626C6F620A,{"another_key":"another_value","key":"value"}
courageous,572,536,-0.134892,2025/07/27,2007/12/28 04:04:33.089,0,626C6F620A,{"another_key":"another_value","key":"value"}
murky,321,3836,-29.019806,2013/09/14,2020/05/20 20:01:18.398,0,626C6F620A,{"another_key":"another_value","key":"value"}
current,946,1786,-75.821284,2024/01/23,2015/06/29 17:44:11.773,1,626C6F620A,{"another_key":"another_value","key":"value"}
bad,906,-4961,58.555764,2021/09/28,2017/05/19 03:07:07.501,0,626C6F620A,{"another_key":"another_value","key":"value"}
fall,655,4850,62.768847,2022/01/25,2025/09/28 19:17:40.143,1,626C6F620A,{"another_key":"another_value","key":"value"}
fall,847,-4739,-189.71033,2013/01/20,2028/10/30 08:16:35.507,1,626C6F620A,{"another_key":"another_value","key":"value"}
scrawny,227,4535,8.927695,2019/06/29,2025/09/12 14:08:51.147,1,626C6F620A,{"another_key":"another_value","key":"value"}
huge,-12,-4563,-160.145695,2007/11/27,2027/11/03 05:32:52.387,0,626C6F620A,{"another_key":"another_value","key":"value"}
courageous,731,-2645,-33.973891,2009/04/14,2013/02/14 11:24:52.293,0,626C6F620A,{"another_key":"another_value","key":"value"}
read,100,-1557,-28.323145,2011/03/13,2013/07/16 09:32:22.022,0,626C6F620A,{"another_key":"another_value","key":"value"}
study,56,-674,-162.842936,2021/02/21,2010/06/28 16:41:23.595,1,626C6F620A,{"another_key":"another_value","key":"value"}
bad,702,-198,140.328749,2022/03/20,2026/04/14 08:38:41.864,1,626C6F620A,{"another_key":"another_value","key":"value"}
current,430,-3908,-157.329372,2013/09/27,2024/12/29 02:45:42.003,0,626C6F620A,{"another_key":"another_value","key":"value"}
huge,895,1686,53.586338,2025/10/17,2018/10/12 13:54:14.202,0,626C6F620A,{"another_key":"another_value","key":"value"}
read,-18,736,41.903661,2025/05/08,2014/02/23 16:35:14.647,1,626C6F620A,{"another_key":"another_value","key":"value"}
legal,397,-684,166.452107,2012/08/19,2020/04/12 16:57:19.989,0,626C6F620A,{"another_key":"another_value","key":"value"}
